### PR TITLE
Tested for pip module in reinstall (#835)

### LINKED ISSUE #835: reinstall-all does not recover from python version upgrade
**Describe the bug**
After the system's python version upgraded (for example, 3.9 -> 3.10), `pipx upgrade-all` fails with `pip` module issue (this is kinda expected):

```
$ pipx upgrade-all
/home/fishy/.local/pipx/shared/bin/python: No module named pip
Failed to upgrade shared libraries
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pipx/shared_libs.py", line 113, in upgrade
    subprocess_post_check(upgrade_process)
  File "/usr/lib/python3/dist-packages/pipx/util.py", line 202, in subprocess_post_check
    raise PipxError(
pipx.util.PipxError: '/home/fishy/.local/pipx/shared/bin/python -m pip --disable-pip-version-check install -q --upgrade pip setuptools wheel' failed
/home/fishy/.local/pipx/venvs/poetry/bin/python: No module named pip
Error encountered when upgrading poetry:
'/home/fishy/.local/pipx/venvs/poetry/bin/python -m pip install --upgrade poetry -q' failed

Versions did not change after running 'pipx upgrade' for each package 😴

Some packages encountered errors during upgrade.
    See specific error messages above.
```

But when I try to recover it from `pipx reinstall-all` (which is usually useful to recover from other minor upgrades), it does not recover, and also uninstall my already installed packages:

```
$ pipx reinstall-all
/home/fishy/.local/pipx/shared/bin/python: No module named pip
Failed to upgrade shared libraries
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pipx/shared_libs.py", line 113, in upgrade
    subprocess_post_check(upgrade_process)
  File "/usr/lib/python3/dist-packages/pipx/util.py", line 202, in subprocess_post_check
    raise PipxError(
pipx.util.PipxError: '/home/fishy/.local/pipx/shared/bin/python -m pip --disable-pip-version-check install -q --upgrade pip setuptools wheel' failed
uninstalled poetry! ✨ 🌟 ✨
Fatal error from pip prevented installation. Full pip output in file:
    /home/fishy/.local/pipx/logs/cmd_2022-04-23_10.34.06_pip_errors.log

Error installing poetry.
The following package(s) failed to reinstall: poetry
$ ls -l .local/pipx/venvs/
total 0
```

I have to remove `~/.local/pipx/shared` and manually reinstall packages to recover.

**How to reproduce**

**Expected behavior**

1. `pipx reinstall-all` really should not remove the packages even if it cannot recover them
2. if there's no easy way to recover automatically, maybe print out a message saying something like "manually remove your ~/.local/pipx/shared might help"?

### PR DESCRIPTION
**Pull request by chrysle**
* [x] I have added an entry to `docs/changelog.md`

## Summary of changes

Closes #835 

## Test plan

Tested by running
```
python3 src/pipx reinstall-all # with poetry installed
nox -s tests-3.8 -- -k test_reinstall
```

### CODE REVIEW

**Comment by uranusjr** [comment]
The message would need some improvements. Simply adding _Fatal error_ is probably not going to be too much more helpful…

**Review by NeoGlitchBob — APPROVED** [review]

**Comment by chrysle** [comment]
@uranusjr Is that better?

**Comment by meowmeowmeowcat** [comment]
I think an error message like ```You might have encountered issues on uninstalling packages. Please remove {PIPX_SHARED_LIBS} and then run `pipx reinstall-all` to fix it.``` would be more helpful, since it's a better solution for users when they encounter `No module named pip` error.

(`PIPX_SHARED_LIBS` can be imported from `pipx.constants`)

**Comment by meowmeowmeowcat** [comment]
Please fix lint errors.

**Comment by chrysle** [comment]
> Please fix lint errors.

Sorry, couldn't run the linter because of missing python interpreter in version 3.11. Now fixed.

File: src/pipx/commands/reinstall.py [hunk 1]
```diff
@@ -44,6 +45,13 @@ def reinstall(
         package_or_url = venv.pipx_metadata.main_package.package_or_url
     else:
         package_or_url = venv.main_package_name
+    
+    try:
+        import pip
+    except ModuleNotFoundError:
+        raise PipxError(
+            f"You might have encountered issues on uninstalling packages. Please remove {PIPX_SHARED_LIBS} and then run `pipx reinstall-all` to fix it."
```
Comment (uranusjr) [inline src/pipx/commands/reinstall.py:53]:
```suggestion
            f"You might have encountered issues on uninstalling packages. "
            f"Remove {PIPX_SHARED_LIBS} and run 'pipx reinstall-all' to fix it."
```

**Comment by uranusjr** [comment]
Can you add a line to the changelog? Otherwise lgtm.

**Pull request merged**